lib/Demangling/Demangler.cpp

```diff
@@ -1330,6 +1330,39 @@ NodePointer Demangler::demangleExtensionContext() {
   return Ext;
 }
 
+/// Associate any \c OpaqueReturnType nodes with the declaration whose opaque
+/// return type they refer back to.
+static Node *setParentForOpaqueReturnTypeNodes(Demangler &D,
+                                              Node *parent,
+                                              Node *visitedNode) {
+  if (!parent || !visitedNode)
+    return nullptr;
+  if (visitedNode->getKind() == Node::Kind::OpaqueReturnType) {
+    // If this node is not already parented, parent it.
+    if (visitedNode->hasChildren()
+        && visitedNode->getLastChild()->getKind() == Node::Kind::OpaqueReturnTypeParent) {
+      return parent;
```

Comment:
Is returning `parent` right in this case? Or should we be returning the parent extracted from the `OpaqueReturnTypeParent` node?

Reply:
This check is a guard against accidentally re-parenting nodes that were already visited. The check below that prevents recursion in function/var/subscript nodes should also prevent this, but since we build the tree bottom-up this should also provide a layer of protection against it. The return value is following the ObjC-ish "return nil if any of your inputs are nil" pattern used elsewhere in the demangler to avoid peppering `if (!node)` checks all over the place; it should only ever be either `parent` or `nullptr`.